storage/src/main/java/tech/pegasys/teku/storage/store/Store.java

```diff
@@ -609,14 +640,43 @@ private SafeFuture<Optional<StateAndBlockSummary>> getOrRegenerateBlockAndState(
     if (cachedResult.isPresent()) {
       return SafeFuture.completedFuture(cachedResult);
     }
+
+    // is it an epoch boundary?
+    if (maybeEpochStates.isPresent()) {
+      final Optional<StateAndBlockSummary> maybeEpochState =
+          maybeEpochStates.get().stream()
+              .filter(
+                  stateAndBlockSummary ->
+                      stateAndBlockSummary.getBlockSummary().getRoot().equals(blockRoot))
+              .findFirst();
```

Comment:
Might be worth making `maybeEpochStates` a map keyed by block root to make lookups faster?